### PR TITLE
Docs: JavaScript tests

### PR DESCRIPTION
**Pull request by annevk**
This tries to pull some of the more pertinent content forward.

<s>TODO:

* create more links into testharness-api 
* figure out how to weave in testdriver.js references
* rewrite idlharness to account for all the changes to how we do that today</s>

Feedback appreciated.

### CODE REVIEW

File: docs/writing-tests/testharness.md
Comment (jgraham) [inline docs/writing-tests/testharness.md:189]: It's a little odd to mention variants here before they're introduced below.

File: docs/writing-tests/testharness.md
Comment (jgraham) [inline docs/writing-tests/testharness.md:212]: s/to be useful/to complete inside the timeout/

**Review by jgraham — COMMENTED** [review]
Generally the testharness.js docs changes lgtm; I didn't read the idlharness ones as closely since others use that more than me,

File: docs/writing-tests/testharness.md
Comment (annevk) [reply docs/writing-tests/testharness.md:189]: Should I make it part of the variants section somehow? It definitely seems better to have variants listed later on.

**Comment by annevk** [comment]
Thanks, I hope @stephenmcgruer can review the IDL part.

File: docs/writing-tests/testharness.md
Comment (jugglinmike) [inline docs/writing-tests/testharness.md:208]: This explanation glosses over an important part of the feature: it doesn't connect the elements to the URL. For instance, "Test runners will execute the test for each 'variant' specified, appending the corresponding `content` value to the URL as they do so." I recognize this part hasn't changed, but would you mind adding something to that effect while we're here?

**Review by jugglinmike — APPROVED** [review]
Looks great!

File: docs/writing-tests/testharness.md
Comment (Hexcles) [inline docs/writing-tests/testharness.md:40]:
```suggestion
If you only need to test a [single thing](https://web-platform-tests.org/writing-tests/testharness-api.html#single-page-tests),
you could also use:
```

File: docs/writing-tests/testharness.md
Comment (Hexcles) [inline docs/writing-tests/testharness.md:24]: I can see how this might be a more logical structure for test authors! As a tooling/infra maintainer, I always think of it as "I write the test page" vs. "server generates the test page" on the top level.

**Review by Hexcles — APPROVED** [review]
Thanks! I like the change overall and just have some nitpicky/optional suggestions.

File: docs/writing-tests/testharness.md
Comment (sideshowbarker) [inline docs/writing-tests/testharness.md:214]:
```suggestion
`/common/subset-tests.js` and `/common/subset-tests-by-key.js` are two utility
scripts that work well together with variants, allowing a test to be split up into
subtests in cases when there are otherwise too many tests to complete inside the
timeout. For example:
```

File: docs/writing-tests/testharness.md
Comment (sideshowbarker) [inline docs/writing-tests/testharness.md:237]:
```suggestion
query string can include or exclude a key (which will be matched as a regular
```

**Review by sideshowbarker — APPROVED** [review]

File: docs/writing-tests/testharness.md
Comment (annevk) [reply docs/writing-tests/testharness.md]: Unfortunately there's a lint that checks things are listed in a ToC.

File: docs/writing-tests/idlharness.md
Comment (annevk) [reply docs/writing-tests/idlharness.md]: I dropped the sentence, it's not very relevant.

File: docs/writing-tests/testharness.md
Comment (annevk) [reply docs/writing-tests/testharness.md:25]: Yeah, server documentation is still suboptimal and might warrant another look to ensure common patterns are covered adequately.

**Comment by annevk** [comment]
I plan on merging this tomorrow. Let me know if you need more time.

File: docs/writing-tests/idlharness.md
Comment (stephenmcgruer) [inline docs/writing-tests/idlharness.md:107]: Nit; it may be worth noting that this is discouraged, and in general IDL should be in an `/interfaces/` file and loaded via the `srcs` parameter?

File: docs/writing-tests/idlharness.md
Comment (stephenmcgruer) [inline docs/writing-tests/idlharness.md:114]: It may be worth being a little bit more explicit about when this is useful. I'm honestly not entirely clear myself, for example `entries-api/idlharness.html` has:

```
idl_array.add_untested_idls('[Exposed=(Window,Worker)] interface File {};');
```

But in its idl file it has:

```
partial interface File {
    readonly attribute USVString webkitRelativePath;
};
```

Does this mean we're not testing for `webkitRelativePath`? Or that in all **other** cases when one uses a partial interface we're incorrectly testing too much?

**Review by stephenmcgruer — APPROVED** [review]
IDL section LGTM with two non-blocking suggestions

**Comment by annevk** [comment]
Thanks @stephenmcgruer for highlighting that. I created #25327, #25331, #25334, #25353, #25354, #25355, and #25356, as well as some PRs that are already merged, to convince myself that we have no need for those methods and everything we have can be done with `idl_test` plus the remaining two methods on `IdlArray`. (There's still some consumers left beyond those PRs but they can all be converted as far as I can tell and I might well do those too.)

I'll give this another day before merging given the changes I just made to tighten up the idlharness documentation.

**Comment by annevk** [comment]
The one instance I forgot about is `idlArray.add_untested_idls('typedef Window WindowProxy;');` in HTML. But let's consider that quirk in IDL itself for not knowing about the Window/WindowProxy interaction.

**Review by stephenmcgruer — APPROVED** [review]
Still LGTM - thanks for going the extra mile here!

**Pull request merged**
